### PR TITLE
Re-enable unused warnings for scala 2.12

### LINKED ISSUE #307: Enabled unused warnings on Scala 2.12
Somehow these fell through the cracks. H/t @satorg

### PR DESCRIPTION
**Pull request by satorg**
Closes #307 (hopefully)

### CODE REVIEW

File: settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala
Comment (satorg) [inline settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala:92]: Why do we remove `-Xlint` for 2.13?

Comment (DavidGregory084) [reply settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala:92]: I don't believe that `-Xlint` on its own exists in 2.13 - I think it was replaced by the multi-argument one `-Xlint:_`, but I could be misremembering

Comment (satorg) [reply settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala:92]: Hmm.. I tried briefly both `-Xlint` and `-Xlint:_` for both Scala 2.12 and 2.13. And seems all these combinations work in a similar way:
```
$ scala-cli compile -j 11 -S 2.12 '-Xlint' test1.scala 
[warn] ./test1.scala:2:31: parameter value ord in method hello is never used
[warn]   def hello[A](a: A)(implicit ord: Ordering[A]): Unit = {
[warn]                               ^
$ scala-cli compile -j 11 -S 2.12 '-Xlint:_' test1.scala 
[warn] ./test1.scala:2:31: parameter value ord in method hello is never used
[warn]   def hello[A](a: A)(implicit ord: Ordering[A]): Unit = {
[warn]                               ^
$ scala-cli compile -j 11 -S 2.13 '-Xlint' test1.scala 
[warn] ./test1.scala:2:31: parameter value ord in method hello is never used
[warn]   def hello[A](a: A)(implicit ord: Ordering[A]): Unit = {
[warn]                               ^^^^^^^^^^^^^^^^
$ scala-cli compile -j 11 -S 2.13 '-Xlint:_' test1.scala 
[warn] ./test1.scala:2:31: parameter value ord in method hello is never used
[warn]   def hello[A](a: A)(implicit ord: Ordering[A]): Unit = {
[warn]                               ^^^^^^^^^^^^^^^^
~
```
Although it is hard to say empirically whether they are fully interchangeable or not.

Comment (satorg) [reply settings/src/main/scala/org/typelevel/sbt/TypelevelSettingsPlugin.scala:92]: Anyway, removing `-Xlint` there is no `-Xlint:_` added for 2.13.

File: github-actions/src/main/scala/org/typelevel/sbt/gha/GenerativePlugin.scala
Comment (satorg) [inline github-actions/src/main/scala/org/typelevel/sbt/gha/GenerativePlugin.scala:617]: unused, so began failing after re-enabling the warnings.

File: core/src/main/scala/org/typelevel/sbt/TypelevelPlugin.scala
Comment (satorg) [inline core/src/main/scala/org/typelevel/sbt/TypelevelPlugin.scala:118]: Unused. Seems it was copied to here:
https://github.com/typelevel/sbt-typelevel/blob/c810fa9f8d3d71c9fc18158537699ac7f62891e1/ci/src/main/scala/org/typelevel/sbt/TypelevelCiPlugin.scala#L129-L132

**Comment by satorg** [comment]
@armanbilge I think it is pretty much ready, could you take a look please?

Btw, there are no "unused" warnings enabled neither for Scala 3.0 nor 3.1 as well.

**Comment by armanbilge** [comment]
Thanks for all your work on this! Much appreciated.

To make up for the lack of tests 😬 I usually try to dog-food things to confirm they are working. Along those lines maybe we can add something like this inside the `TypelevelSettingsPlugin`:
```scala
@nowarn("cat=unused")
private[this] def unused: Unit = ()
```

Also, as usual I am terrified what will happen when we roll this out to 100+ projects 😅 but assuming that nearly all projects cross-building for 2.12 are also building for 2.13, and there are no weird differences in unused warnings, it should be okay.

> Btw, there are no "unused" warnings enabled neither for Scala 3.0 nor 3.1 as well.

Unfortunately that's because nobody has implemented them in dotty yet 😕

File: core/src/main/scala/org/typelevel/sbt/TypelevelPlugin.scala
Comment (armanbilge) [reply core/src/main/scala/org/typelevel/sbt/TypelevelPlugin.scala:118]: Nice!

File: github-actions/src/main/scala/org/typelevel/sbt/gha/GenerativePlugin.scala
Comment (armanbilge) [reply github-actions/src/main/scala/org/typelevel/sbt/gha/GenerativePlugin.scala:617]: Also nice!

File: mergify/src/main/scala/org/typelevel/sbt/mergify/MergifyAction.scala
Comment (armanbilge) [inline mergify/src/main/scala/org/typelevel/sbt/mergify/MergifyAction.scala:61]: Cool, seems we basically already have the test I'm looking for 😂

**Comment by satorg** [comment]
> Along those lines maybe we can add something like this inside the TypelevelSettingsPlugin:
```
@nowarn("cat=unused")
private[this] def unused: Unit = ()
```
In fact, I actually added `@nowarn("cat=unused")` in a couple of places already, because they began failing without it.
So we can say for sure that these changes are working to some extent. 

Do you think it make sense to add one more such an annotation directly just for testing?

**Comment by satorg** [comment]
> Unfortunately that's because nobody has implemented them in dotty yet 😕

Yes, you're right. Although Scala 3.1 has added support for `-Wunused:all` but seems it does nothing.

**Comment by armanbilge** [comment]
> Do you think it make sense to add one more such an annotation directly just for testing?

Sorry, my comments were stuck in pending 😆 yes, I did notice those :) I think a deliberate one in the settings plugin would be good.

**Review by armanbilge — APPROVED** [review]
👍 in principle, I don't really understand scalac options 😂  getting @DavidGregory084's sign-off would be great.

I might let this live in a snapshot for a bit, which should be enough to unblock the Cats work and will give me a chance to make sure this doesn't cause any surprises in e.g. http4s.

**Comment by satorg** [comment]
Hmm.. I'm struggling to understand, why has the workflow failed?

**Comment by armanbilge** [comment]
Try running `scalafixAll` ?

**Comment by satorg** [comment]
I did... It succeeds if I run it locally.

**Comment by armanbilge** [comment]
It succeeds, but then makes some changes to commit? :)

**Comment by armanbilge** [comment]
Wooops, I just realized this was targeted at main. If we want to get this into the next 0.4.x release it should be targeted at `series/0.4`.

**Comment by satorg** [comment]
Hmm... Seems that somehow I missed your last comment @armanbilge, sorry. Does it still make sense to backport these changes to `series/0.4`?

**Comment by armanbilge** [comment]
No problem. I was always on the fence about this, since 0.5.x is pretty close (we'll need to bump for sbt 1.7.0 and laika 0.19.0) and I wasn't sure how this will affect some builds in case 2.12 has different warnings for some reason.

That is to say, I think you needed this for you Cats PR, so we can put you on a snapshot while you work on it and then something more stable when its ready to merge.

**Pull request merged**
